### PR TITLE
Add custom callback support for DDP training

### LINKED ISSUE #6168: Custom callback functions doesn't called in DDP training
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Multi-GPU

### Bug

When I add custom callbacks in YOLOv8, then they are only called when I train with one GPU. As soon as I train with multiple GPUs, the callback functions are not being called anymore.

I see similiar issues in [here](https://github.com/ultralytics/ultralytics/issues/1072), but custom callback function still doesn't applied to me. Ultralytics package is newest version(YOLOv8.0.207) as i know.

### Environment

```
Ultralytics YOLOv8.0.207 🚀 Python-3.9.18 torch-1.10.1+cu111 CUDA:0 (NVIDIA TITAN RTX, 24220MiB)
Setup complete ✅ (20 CPUs, 125.5 GB RAM, 319.1/372.3 GB disk)

OS                  Linux-5.4.0-150-generic-x86_64-with-glibc2.27
Environment         Linux
Python              3.9.18
Install             pip
RAM                 125.49 GB
CPU                 Intel Core(TM) i9-10900X 3.70GHz
CUDA                11.1

matplotlib          ✅ 3.8.1>=3.3.0
numpy               ✅ 1.26.1>=1.22.2
opencv-python       ✅ 4.8.1.78>=4.6.0
pillow              ✅ 10.1.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.3>=1.4.1
torch               ✅ 1.10.1+cu111>=1.8.0
torchvision         ✅ 0.11.2+cu111>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
pandas              ✅ 2.1.2>=1.1.4
seaborn             ✅ 0.13.0>=0.11.0
psutil              ✅ 5.9.6
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1

```

### Minimal Reproducible Example

```
from ultralytics import YOLO

def on_pretrain_routine_end(trainer):
    print("\n>>>>>>>> YOLO CALLBACK >>>>>>>>\n")

def on_fit_epoch_end(trainer):
    print("\n>>>>>>>> YOLO CALLBACK >>>>>>>>\n")

def on_train_end(trainer):
    print("\n>>>>>>>> YOLO CALLBACK >>>>>>>>\n")

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
model.add_callback("on_pretrain_routine_end", on_pretrain_routine_end)
model.add_callback("on_fit_epoch_end", on_fit_epoch_end)
model.add_callback("on_train_end", on_train_end)

# Use the model
model.train(data="coco128.yaml", epochs=3, device="0,1")  # with device="0" it works
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #18070: Custom Callbacks for ClearML Failing During Multi-GPU Training
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When executing training tasks on multiple GPUs, the custom callbacks integrated with ClearML are not functioning as expected. This issue seems to arise specifically during multi-GPU operations, while single GPU training does not exhibit the same problem. When setting `device=0`, the callbacks function correctly. However, if you set `device=0,1,2,...`, the callbacks do not work. Perhaps such behavior of callbacks without ClearML is the same. The training process is executed by the ClearML agent on a computing server during the ClearML pipeline.

A similar issue has been discussed here [#6913](https://github.com/ultralytics/ultralytics/issues/6913). Although adding a RANK check to the code seemed to help in the past, it no longer resolves the issue.

### Environment

Ultralytics 8.3.0 🚀 Python-3.12.3 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3070 Ti, 7869MiB)
Setup complete ✅ (16 CPUs, 31.3 GB RAM, 51.1/93.2 GB disk)

OS                  Linux-6.8.0-49-generic-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             pip
RAM                 31.27 GB
CPU                 AMD Ryzen 7 5800X 8-Core Processor
CUDA                12.1

numpy               ✅ 1.26.4<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1>=1.8.0
torchvision         ✅ 0.19.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.6>=2.0.0
torch               ✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"

### Minimal Reproducible Example

```
from clearml import Task
from ultralytics import YOLO
from ultralytics.utils import RANK


def checkpoint_save_callback(trainer):
    if RANK in {-1, 0}:
        if task := Task.current_task():
            task.update_output_model(
                model_path=str(trainer.last),
                model_name="last",
                auto_delete_file=False,
                iteration=trainer.epoch,
            )
            if trainer.save_period > 0 and trainer.epoch % trainer.save_period == 0:
                model_name = f"epoch{trainer.epoch}"
                task.update_output_model(
                    model_path=str(trainer.last.with_stem(model_name)),
                    model_name=model_name,
                    auto_delete_file=False,
                    iteration=trainer.epoch,
                )


def train(model_path: str, config: dict) -> None:
    model = YOLO(model_path)
    if config.get("save_period", -1) > 0:
        model.add_callback("on_model_save", checkpoint_save_callback)
    model.train(**config)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/18070

Custom callbacks don't work with DDP training since callbacks can't be passed directly. This PR tries to use `dill` to dump the callbacks to a pickle file and load it in the DDP process.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improves support for custom callbacks in distributed (DDP) training and cleans up unused code for a smoother training experience. 🚀

### 📊 Key Changes
- Adds serialization and sharing of custom callbacks when using DDP (Distributed Data Parallel) training, ensuring all processes use the same callbacks.
- Removes unnecessary code that deleted the training save directory during DDP setup.

### 🎯 Purpose & Impact
- Ensures custom callbacks work seamlessly in multi-GPU (DDP) training, making advanced workflows more reliable and flexible.
- Prevents accidental loss of training outputs by no longer deleting the save directory.
- Enhances stability and user experience for those training large models or using custom training hooks.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features, particularly for DDP-related changes or the new callback serialization process.
- ✅ **Add Tests**: Include or update tests to validate callback serialization/deserialization and to verify that the `save_dir` logic functions as intended. Confirm all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Ensure your changes are limited to the **minimum** necessary for adding the custom callback serialization and updating `save_dir` logic. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing).  

💡 Additionally, if you encounter edge cases or potential issues with specific configurations (e.g., environments where `dill` is unavailable), it would be helpful to explicitly document these considerations or provide relevant tests.

Thank you for enhancing Ultralytics' distributed training workflow! 🚀 An Ultralytics engineer will review your PR and provide further feedback shortly.

**Comment by holladayinnn** [comment]
What is around the estimated time this will get merged into main?

**Comment by Y-T-G** [comment]
@holladayinnn You can upvote it. It would only be merged if there's sufficient interest.
